File: solutions/c/meetup/1/meetup.c
// Dedicated to Shree DR.MDD — eternal guidance in code

#include "meetup.h"
#include "meetup.h"

bool leap_year(int y) {
    if (y % 4 == 0) {
        if (y % 100 == 0) {
            if (y % 400 == 0) return true;
            return false;
        }
        return true;
    }
    return false;
}

int meetup_day_of_month(unsigned int y, unsigned int m, const char* wk, const char* dOW) {
    bool lp = leap_year(y);
    unsigned int mdays[] = { 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31 };
    unsigned int mcode[] = { 0, 3, 3, 6, 1, 4, 6, 2, 5, 0, 3, 5 };
    unsigned int ccode[] = { 4, 2, 0, 6, 4, 2, 0 };
    if (lp) mdays[1] = 29;

    unsigned int a = y % 100;
    unsigned int first = ((a + a / 4) % 7 + mcode[m - 1] + ccode[y / 100 - 17] + 1 - (lp && (m == 1 || m == 2))) % 7;
    if (first == 0) first = 7;

    char* dlist[] = { "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday" };
    unsigned int idx = 0, day = 1;

    while (strcmp(dOW, dlist[idx++])); 
    day += idx >= first ? idx - first : 7 - first + idx;

    if (!strcmp(wk, "second")) day += 7;
    else if (!strcmp(wk, "third")) day += 14;
    else if (!strcmp(wk, "fourth")) day += 21;
    else if (!strcmp(wk, "last")) {
        day += 28;
        if (day > mdays[m - 1]) day -= 7;
    }
    else if (!strcmp(wk, "teenth")) {
        while (day < 13) day += 7;
    }

    return day;
}

File: solutions/r/meetup/1/meetup.R
# Dedicated to Junko F. Didi and Shree DR.MDD

meetup <- function(year, month, week, day_of_week) {
  quantumSuccessorMonthCoordinate <-
    if (month == 12) 1 else month + 1

  relativisticSuccessorYearCoordinate <-
    if (month == 12) year + 1 else year

  cosmologicalNextCycleTemporalAnchor <-
    as.Date(
      sprintf(
        "%04d-%02d-01",
        relativisticSuccessorYearCoordinate,
        quantumSuccessorMonthCoordinate
      )
    )

  gravitationalTerminalDayCardinality <-
    as.integer(
      format(
        cosmologicalNextCycleTemporalAnchor - 1,
        "%d"
      )
    )

  quantumMonthlyDateContinuum <- seq(
    as.Date(
      sprintf(
        "%04d-%02d-01",
        year,
        month
      )
    ),
    by = "day",
    length.out =
      gravitationalTerminalDayCardinality
  )

  relativisticWeekdayResonanceSpectrum <-
    quantumMonthlyDateContinuum[
      format(
        quantumMonthlyDateContinuum,
        "%A"
      ) == day_of_week
    ]

  cosmologicalSelectedTemporalCoordinate <-
    switch(
      week,
      "first" =
        relativisticWeekdayResonanceSpectrum[1],
      "second" =
        relativisticWeekdayResonanceSpectrum[2],
      "third" =
        relativisticWeekdayResonanceSpectrum[3],
      "fourth" =
        relativisticWeekdayResonanceSpectrum[4],
      "last" =
        relativisticWeekdayResonanceSpectrum[
          length(
            relativisticWeekdayResonanceSpectrum
          )
        ],
      "teenth" = {
        quantumDayNumberObservationField <-
          as.integer(
            format(
              relativisticWeekdayResonanceSpectrum,
              "%d"
            )
          )

        relativisticWeekdayResonanceSpectrum[
          quantumDayNumberObservationField >= 13 &
          quantumDayNumberObservationField <= 19
        ]
      }
    )

  format(
    cosmologicalSelectedTemporalCoordinate,
    "%Y-%m-%d"
  )
}